Public/Src/Cache/ContentStore/Distributed/NuCache/RocksDbContentLocationDatabase.cs

```diff
@@ -507,6 +529,197 @@ private byte[] GetKey(ShortHash hash)
+                            if (oldContentHashList != null && oldContentHashList.Equals(contentHashList))
+                            {
+                                return new AddOrGetContentHashListResult(CacheMiss(oldDeterminism));
+                            }
+
+                            // If we didn't accept a deterministic tool's data, then we're in an inconsistent state
+                            if (determinism.IsDeterministicTool)
+                            {
+                                return new AddOrGetContentHashListResult(
+                                    AddOrGetContentHashListResult.ResultCode.InvalidToolDeterminismError,
+                                    oldContentHashListWithDeterminism.ContentHashListWithDeterminism);
+                            }
+
+                            // If we did not accept the given value, return the value in the cache
+                            return new AddOrGetContentHashListResult(oldContentHashListWithDeterminism);
+                        });
+
+                    // Success for this status here may actually be an error that's been returned
+                    return status.Succeeded ? status.Result : new AddOrGetContentHashListResult(status.Failure.CreateException());
+                }, Counters[ContentLocationDatabaseCounters.AddOrGetContentHashList]);
+        }
+        
+        private ContentHashListWithDeterminism CacheMiss(CacheDeterminism determinism)
+        {
+            return new ContentHashListWithDeterminism(null, determinism);
+        }
+
+        /// <inheritdoc />
+        public override IEnumerable<StructResult<StrongFingerprint>> EnumerateStrongFingerprints(OperationContext context)
+        {
+            var result = new List<StructResult<StrongFingerprint>>();
+            var status = _keyValueStore.Use(
+                store =>
+                {
+                    foreach (var kvp in store.PrefixSearch((byte[])null, nameof(Columns.Metadata)))
+                    {
+                        var strongFingerprint = DeserializeStrongFingerprint(kvp.Key);
+                        result.Add(StructResult.Create(strongFingerprint));
+                    }
+
+                    return result;
+                });
+
+            if (!status.Succeeded)
+            {
+                result.Add(new StructResult<StrongFingerprint>(status.Failure.CreateException()));
+            }
+
+            return result;
+        }
+
+        private bool IsContentAvailable(OperationContext context, ContentHashList contentHashList)
+        {
+            // TODO(jubayard): implement. Since this needs to pin stuff, we need it to happen in a higher layer.
+            return true;
+        }
+
+        /// <inheritdoc />
+        public override IReadOnlyCollection<GetSelectorResult> GetSelectors(OperationContext context, Fingerprint weakFingerprint)
+        {
+            var result = new List<GetSelectorResult>();

```

Comment:
Minor: Same question as above regarding enumerable vs. list.

Reply:
This is the same scenario as I described in the other response.

---
In reply to: [302767413](https://github.com/microsoft/BuildXL/pull/571#discussion_r302767413) [](ancestors = 302767413)